### DISCUSSION TITLE
Difficulty Index use-case for WPC

### DISCUSSION THREAD

**Discussion by hertneky** [dtcenter/METplus#1755] [Configuration] [answered]
1) The example Difficulty Index (DI) use-case uses a user script to read in a preprocessed data array of ensemble data in dimensions num_lats x num_lons x num_members, computes ensemble mean/std, and then computes DI.

For the WPC use-case, we were looking into instead using gen_ens_prod for computing the mean/std, so that met output file would need to be read in as well as the individual ens member data. So the user script would read in:
- gen_ens_prod output file
- individual member data

The USER_SCRIPT_INPUT_DIR/_TEMPLATE can only read one of those items in as far as I can tell, so I created a user_env_vars to read in the 2nd dataset. My use-case runs fine, but thought I would get input from the metplus team if there was a better/ more ideal way to do this.

2) The data I am using is on a LC grid where dx != dy.

results in the obvious warning:

read_grib2_grid() -> MET does not currently support Lambert Conformal grids where dx (20.3176) != dy (20.4715) and may produce unexpected results!

The met output file shows d_km = 20.3176, so assuming it forced dx=dy? nx and ny are the same as the origin file. It would be good to understand what exactly it does here.

**Comment by georgemccabe** [comment, marked as answer]
Hi @hertneky, you can provide a comma-separated list of templates for USER_SCRIPT_INPUT_TEMPLATE and it will set an environment variable with the path to the file (or text file that contains a list of files if using wildcards) that can be referenced in the user script. By default the environment variables are named METPLUS_FILELIST_INPUT<n> (i.e. METPLUS_FILELIST_INPUT0) unless you set USER_SCRIPT_INPUT_TEMPLATE_LABELS where each label will replace INPUT<n> in the environment variable name. Let me know if that does not work or if you need help setting the variables to get it to work.

@JohnHalleyGotway, can you answer the 2nd question about dx/dy?

**Reply by hertneky** [reply]
@georgemccabe Thanks. It does look like, even for a single file, it adds it to a file list and for my single file file_list, the path is correct. For the wildcard template, my file_list is empty (labeled missing in the file list). I don't see any errors in the template. For reference, here is my template:
USER_SCRIPT_INPUT_DIR = {OUTPUT_BASE}
USER_SCRIPT_INPUT_TEMPLATE = pcp_combine/{init?fmt=%Y%m%d%H}/*{lead?fmt=%3H}.nc,gen_ens_prod/{init?fmt=%Y%m%d%H}/gen_ens_prod_W24I_{init?fmt=%Y%m%d%H}_f{lead?fmt=%3H}.nc

filelist0 (no list)
file_list
missing

filelist1 (correct)
file_list
/d1/personal/hertneky/metprd/gen_ens_prod/2020122412/gen_ens_prod_W24I_2020122412_f192.nc

**Reply by georgemccabe** [reply]
@hertneky could you send me your log file so I can take a look? If you don't have LOG_LEVEL=DEBUG, that would be useful to see what it is trying to search.

**Reply by georgemccabe** [reply]
Hi @hertneky, just following up on this. Were you able to resolve this issue? If not, could you send a log file with debug output turned on?

**Reply by hertneky** [reply]
@georgemccabe For 1) I did not get it to work as you suggested. I used a work-around where I set the USER_SCRIPT_INPUT_TEMPLATE to list the gen_ens_prod input file and created a user_env_vars variable to list the input ensemble member files.
With your suggestion to list both in USER_SCRIPT_INPUT_TEMPLATE, please see the logfile (ERROR does not support multiple files) /d1/personal/hertneky/metprd/logs/metplus.log.20220907162519

2) Was not answered yet.

**Reply by JohnHalleyGotway** [reply]
@hertneky, yes, the MET grid library code does not support lambert conformal grids where dx != dy. Here is the [line of code](https://github.com/dtcenter/MET/blob/main_v10.1/met/src/libcode/vx_data2d_grib2/data2d_grib2.cc#L1230) where that warning message is printed. And here is the [line of code](https://github.com/dtcenter/MET/blob/e213c501c932aa0a26730a87224c418876ecd2d0/met/src/libcode/vx_data2d_grib2/data2d_grib2.cc#L1221) a few above that, where the code is setting the "d_km" value using dx, not dy.

@rgbullock is the best resource for details on this. I believe that dx != dy could theoretically be supported but would require significant changes to the MET grid library.

If actually supporting dx != dy is a requirement for some project, we'll need to clearly define it and decide if/how to pay for that enhancement. Note that the same limitation exists for Polar Stereographic grids, as seen on this [line of code](https://github.com/dtcenter/MET/blob/e213c501c932aa0a26730a87224c418876ecd2d0/met/src/libcode/vx_data2d_grib2/data2d_grib2.cc#L1163).

**Reply by hertneky** [reply]
Thanks @JohnHalleyGotway, this is all the info I needed. I don't think support for this is necessary at this time. I have a feeling this issue emerged from the WPC super ensemble (WSE) data processing. It sounded like the WSE data format was grads and they converted to grib2.

**Reply by hertneky** [reply]
@JohnHalleyGotway correction, the WSE data are originally in **gempak** format, and confirmed that the dx/dy are also not equal in those files. Mike is going to contact Bruce about this just to see what he says, but I am not sure if it's something they have control over. This is the first time I have ever seen an LC grid with unequal dx/dy, which still I believe does not justify adding to METplus.

**Comment by georgemccabe** [comment]
It looks like there is a bug that is preventing UserScript from running this way using wildcard characters. I can provide a version of METplus for you to test if the change fixes this issue. I will find someone to answer the dx/dy question, as I don't know what the code is actually doing in this case.

**Reply by hertneky** [reply]
@georgemccabe Thanks George, this sounds good. I can test this with this case when ready. @JohnHalleyGotway answered the other question - which was just understanding what the code was doing in this case and how d_km was set.

**Comment by georgemccabe** [comment]
@hertneky, thanks for testing the change. I merged it into the develop branch so it will be available in the next beta release. Let me know if there is anything else you need for this ticket.
